hvac/vitobloc/firmware.lua

```diff
@@ -0,0 +1,369 @@
+    end
+    if value & 2 then
+      table.insert(external_errors, 'power_module_reverse')
+    end
+    return external_errors
+  end
+end
+
+function parse_other_error(value)
+  if not value then
+    return {}
+  end
+  if value == 0 then
+    return { 'unknown_value' }
+  end
+  if type(value) == 'number' then
+    local other_errors = {}
+
+    if value & 1 then
+      table.insert(other_errors, 'pump_dry_1')
+    end
+    if value & 2 then
+      table.insert(other_errors, 'pump_dry_2')
+    end
+    if value & 4 then
+      table.insert(other_errors, 'pump_dry')
+    end
+    return other_errors
+  end
+end
+
+---------------------------------
+-- Vitobloc ModbusTCP API
+---------------------------------
+
+VitoblocModbusTcp = {}
+
+function VitoblocModbusTcp.new(ip_address, unit_id)
+  assert(
+    type(ip_address) == 'string',
+    'ip_address (arg #1) must be string, given: ' .. inspect(ip_address)
+  )
+  assert(type(unit_id) == 'number', 'unit_id (arg #2) must be number, given: ' .. inspect(unit_id))
+
+  local self = setmetatable({}, { __index = VitoblocModbusTcp })
+  self.ip_address = ip_address
+  self.unit_id = unit_id
+  return self
+end
+
+function VitoblocModbusTcp:connect()
+  self.modbus = modbustcp.new(self.ip_address)
+end
+
+function VitoblocModbusTcp:read_holdings(address, registers_count)
+  assert(type(address) == 'number', 'address (arg #1) must be number, given: ' .. inspect(address))
+  assert(
+    type(registers_count) == 'number',
+    'refisters_count (arg #2) must be number, given: ' .. inspect(registers_count)
+  )
+
+  local registers, err = self.modbus:read_holdings(self.unit_id, address, registers_count, 1000)
+  if err and err ~= 0 then
+    enapter.log('read error: ' .. err, 'error')
+    if err == 1 then
+      -- Sometimes timeout happens and it may break underlying Modbus client,
+      -- this is a temporary workaround which manually reconnects.
+      self:connect()
+    end
+    return nil
+  end
+  return registers
+end
+
+function VitoblocModbusTcp:read_u32(address, factor)
+  if not factor then
+    factor = 1
+  end
+
+  local reg = self:read_holdings(address, 2)
+  if not reg then
+    return
+  end
+
+  local raw = string.pack('>I2I2', reg[1], reg[2])
+  return string.unpack('>I4', raw) * factor
+end
+
+function VitoblocModbusTcp:read_u8(address, factor)
+  if not factor then
+    factor = 1
+  end
+
+  local reg = self:read_holdings(address, 1)
+  if not reg then
+    return
+  end
+
+  -- NaN for U16 values
+  if reg[1] == 0xFFFF then
+    return nil
+  end
+
+  local raw = string.pack('>I2', reg[1])
+  return string.unpack('>I2', raw) * factor
+end
+
+function VitoblocModbusTcp:read_u16(address, factor)
+  if not factor then
+    factor = 1
+  end
+
+  local reg = self:read_holdings(address, 1)
+  if not reg then
+    return
+  end
+
+  -- NaN for U16 values
+  if reg[1] == 0xFFFF and reg[2] == 0xFFFF then
+    return nil
+  end
```

Comment:
one Modbus register = 16 bits. here you're trying to check 2 registers but you'll never get them.
I suggest to delete all of this.